lib/temporal/worker/activity_worker.rb

```diff
@@ -6,44 +6,46 @@ module Temporal
   class Worker
     class ActivityWorker
       def initialize(task_queue, core_worker, activities, converter, executor)
-        @running = false
         @task_queue = task_queue
         @worker = SyncWorker.new(core_worker)
         @activities = prepare_activities(activities)
         @converter = converter
         @executor = executor
         @running_activities = {}
         @cancellations = []
+        @shutdown_queue = Queue.new
       end
 
       def run(reactor)
-        @running = true
+        # @type var async_tasks: Array[Async::Task]
+        async_tasks = []
 
-        while running?
+        loop do
           activity_task = worker.poll_activity_task
-          reactor.async do
+          async_tasks << reactor.async do |async_task|
             if activity_task.start
               handle_start_activity(activity_task.task_token, activity_task.start)
             elsif activity_task.cancel
               handle_cancel_activity(activity_task.task_token, activity_task.cancel)
             end
+
+            async_tasks.delete(async_task)
           end
         end
+      rescue Temporal::Bridge::Error::WorkerShutdown
+        async_tasks.each(&:wait) # wait for all outstanding tasks to finish
+        shutdown_queue << nil
       end
 
       def shutdown
```

Comment:
I assume handling signal interruptions will come in the future?

Reply:
Yes, the plan is to provide a higher level `Worker.run` (as opposed to `Worker#run`) interface that takes a number of workers and runs all of them within the same reactor. It will provide the signal handling as well, while the current `Worker#run` will serve as a bare minimum ability to run a single worker (which some people might still find useful if they have their own execution environment and want to tweak things)